src/components/views/context_menus/MessageContextMenu.tsx

```diff
@@ -221,12 +248,71 @@ export default class MessageContextMenu extends React.Component<IProps, IState>
         this.closeMenu();
     };
 
+    private onCopyPermalinkClick = (e: ButtonEvent): void => {
+        e.preventDefault(); // So that we don't open the permalink
+        copyPlaintext(this.getPermalink());
+        this.closeMenu();
+    };
+
     private onCollapseReplyThreadClick = (): void => {
         this.props.collapseReplyThread();
         this.closeMenu();
     };
 
-    private getReactions(filter: (e: MatrixEvent) => boolean): MatrixEvent[] {
+    private onCopyClick = (): void => {
+        copyPlaintext(this.getSelectedText());
+        this.closeMenu();
+    };
+
+    private onSaveClick = async (): Promise<void> => {
+        const content = this.props.mxEvent.getContent() as IMediaEventContent;
+        const media = mediaFromContent(this.props.mxEvent.getContent());
+
+        if (content.file !== undefined && !this.decryptedUrl) {
+            try {
+                const blob = await decryptFile(content.file);
+                this.decryptedUrl = URL.createObjectURL(blob);
+            } catch (error) {
+                console.warn("Unable to decrypt attachment: ", error);
+                return;
+            }
+        }
+
+        const anchor = document.createElement("a");
+        anchor.href = this.decryptedUrl || media.srcHttp;
+        anchor.download = content.body?.length > 0 ? content.body : _t('Attachment');
+        anchor.target = "_blank";
+        anchor.rel = "noreferrer noopener";
+        anchor.click();
+        this.closeMenu();
+    };
```

Comment:
this will need to use the new `mediaEventHelper` stuff, for caching purposes. Also, we can't safely have the anchor within the scope here: it would need to be done like the action bar download button (which probably means factoring it out from there into somewhere both places can use it)

Reply:
In the end, I decided to drop the download option, since we already use a native menu for images and I am not entirely sure what the plan is with the download functionality. I am happy to reimplement it properly if design wants me to do so